components/uniset/src/uniset.rs

```diff
@@ -124,25 +125,52 @@ impl UnicodeSet {
         &self.inv_list
     }
 
-    /// Yields an iterator going through the character set in the [`UnicodeSet`]
+    /// Yields an [`Iterator`] going through the character set in the [`UnicodeSet`]
     ///
     /// # Examples
     ///
     /// ```
     /// use icu::uniset::UnicodeSet;
     /// let example_list = vec![65, 68, 69, 70];
     /// let example = UnicodeSet::from_inversion_list(example_list).unwrap();
-    /// let mut example_iter = example.iter_chars();
-    /// assert_eq!(Some('A'), example_iter.next());
-    /// assert_eq!(Some('B'), example_iter.next());
-    /// assert_eq!(Some('C'), example_iter.next());
-    /// assert_eq!(Some('E'), example_iter.next());
-    /// assert_eq!(None, example_iter.next());
+    /// let mut ex_iter_chars = example.iter_chars();
+    /// assert_eq!(Some('A'), ex_iter_chars.next());
+    /// assert_eq!(Some('B'), ex_iter_chars.next());
+    /// assert_eq!(Some('C'), ex_iter_chars.next());
+    /// assert_eq!(Some('E'), ex_iter_chars.next());
+    /// assert_eq!(None, ex_iter_chars.next());
     /// ```
     pub fn iter_chars(&self) -> impl Iterator<Item = char> + '_ {
         self.inv_list.chunks(2).flat_map(|pair| (pair[0]..pair[1])).filter_map(char::from_u32)
     }
 
+    /// Yields an [`Iterator`] returning the ranges of the code points that are
+    /// included in the [`UnicodeSet`]
+    ///
+    /// Ranges are returned as [`RangeInclusive`], which is inclusive of its
+    /// `end` bound value. An end-inclusive behavior matches the ICU4C/J
+    /// behavior of ranges, ex: `UnicodeSet::contains(UChar32 start, UChar32 end)`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu::uniset::UnicodeSet;
+    /// let example_list = vec![65, 68, 69, 70];
+    /// let example = UnicodeSet::from_inversion_list(example_list).unwrap();
+    /// let mut example_iter_ranges = example.iter_ranges();
+    /// assert_eq!(Some(65..=67), example_iter_ranges.next());
+    /// assert_eq!(Some(69..=69), example_iter_ranges.next());
+    /// assert_eq!(None, example_iter_ranges.next());
+    /// ```
+    pub fn iter_ranges(&self) -> impl ExactSizeIterator<Item = RangeInclusive<u32>> + '_ {
+        self.inv_list.chunks(2).map(|pair| RangeInclusive::new(pair[0], pair[1] - 1))
```

Comment:
`stylistic, non-blocking`: you can destruct in closure: `|(low, high)| RangeInclusive::new(low, high - 1)`

Reply:
It's an array, not a tuple. `[low, high]` does work as a pattern (might be unstable) but only in fallible contexts since the length could be incorrect.

Reply:
Just a question, I don't know Rust conventions...

It *looks like* this creates a list of RangeInclusive items, effectively copying everything in the inv_list. This works fine, of course, but wouldn't it be possible to implement the iterator by referencing the original inv_list and each time reading the next range from there on the fly? Speed should be the same, but you would save allocating the list and its items.

Or is that what the ExactSizeIterator and `inv_list.chunks(2)` do?

Reply:
Copying ints is extremely cheap and we'll have to do it anyway when we migrate UnicodeSet to ZeroMap.

Reply:
Oh, to be clear, we are not allocating any memory. .chunks and .map read the list iteratively.